gql_error_link/lib/gql_error_link.dart

```diff
@@ -0,0 +1,79 @@
+/// GQL Link to handle execution errors and exceptions
+library gql_error_link;
+
+import "dart:async";
+import "package:async/async.dart";
+import "package:gql_link/gql_link.dart";
+import "package:gql_exec/gql_exec.dart";
+
+/// A handler of GraphQL errors.
+typedef ErrorHandler = Stream<Response> Function(
+  Request request,
+  NextLink forward,
+  Response response,
+);
+
+/// A handler of Link Exceptions.
+typedef ExceptionHandler = Stream<Response> Function(
+  Request request,
+  NextLink forward,
+  LinkException exception,
+);
+
+/// [ErrorLink] allows interception of GraphQL errors (using [onError])
+/// and [LinkException]s (using [onException]).
+///
+/// In both cases [ErrorLink] transfers control over to the handler which may
+/// return a new stream to discard the original stream. If the handler returns
+/// `null`, the original stream is left intact and will be allowed to continue
+/// streaming new events.
+class ErrorLink extends Link {
+  final ErrorHandler onError;
+  final ExceptionHandler onException;
+
+  const ErrorLink({
+    this.onError,
+    this.onException,
+  });
+
+  @override
+  Stream<Response> request(
+    Request request, [
+    forward,
+  ]) async* {
+    await for (final result in Result.captureStream(forward(request))) {
+      if (result.isError) {
+        final error = result.asError.error;
+
+        if (onException != null && error is LinkException) {
+          final stream = onException(request, forward, error);
+
+          if (stream != null) {
+            yield* stream;
+
+            return;
+          }
+        }
+
+        yield* Stream.error(error);
```

Comment:
Will this still allow the link to receive data events after an error?

I'm still not clear on why we wouldn't just use StreamTransformers. It seems like async generators with Result is actually making it more complicated, not less.

Reply:
> Will this still allow the link to receive data events after an error?

Yes. This yields an error without closing the generator.

> I'm still not clear on why we wouldn't just use StreamTransformers.

`StreamTransformer.fromHandlers` appears to be synchronous so it's not possible to redirect the stream and guarantee the `sink` will not be closed before all events are consumed. (Which is a requirement for AuthLink).
If we follow Apollo, Error Link is not a transformer at all. It's a way to get notified about errors and possibly re-sourcing the stream.

Reply:
Ok, that makes sense.